### ISSUE TITLE
Memory leak in GridLayout on Page Navigation

### ISSUE THREAD

**Issue by shripalsoni04** [NativeScript/NativeScript#2712]
### Did you verify this is a real problem by searching [Stack Overflow]
Yes

### Tell us about the problem
When we write code as shown below, and navigate to the page having this code, the ImageView is not getting cleared from heap.
```
<GridLayout>
    <Image src="https://unsplash.it/400/200?image=1067"></Image>
</GridLayout>
```

Instead of using GridLayout, when we use StackLayout, the ImageView is getting cleared properly.
```
<StackLayout>
    <Image src="https://unsplash.it/400/200?image=1067"></Image>
</StackLayout>
```

Similarly when we write code with multiple views as shown below, the views which are inside StackLayout are getting removed but as `<Label row="1">` is a direct child of GridLayout, it is not getting cleared on GC.
```
<GridLayout rows="auto, auto, auto">
    <StackLayout row="0">
        <Label text="Label 1"></Label> <!-- Removed on GC -->
    </StackLayout>
    <Label row="1" text="Label 2"></Label> <!-- Not Removed on GC -->
    <StackLayout row="2">
        <Image src="https://unsplash.it/400/200?image=1067"></Image> <!-- Removed on GC -->
    </StackLayout>
</GridLayout>
```

Just wrapping above code inside StackLayout, clears everything on GC as shown below.
```
<StackLayout>
    <GridLayout rows="auto, auto, auto">
        <StackLayout row="0">
            <Label text="Label 1"></Label>
        </StackLayout>
        <Label row="1" text="Label 2"></Label>
        <StackLayout row="2">
            <Image src="https://unsplash.it/400/200?image=1067"></Image>
        </StackLayout>
    </GridLayout>
</StackLayout>
```

After analysing heapDumps, I came to know that, when we use GridLayout and when we are navigated from the page, the GridLayout keeps all of direct children reference in `helper` variable of GridLayout. So when FinalizerThread runs, it cannot collect the child elements. While we use StackLayout, as may be it is removing its child reference, the FinalizerThread can collect all such views. I may be wrong but I think when new page is opened, framework is wrapping the whole code in GridLayout and as GridLayout is keeping the references, all the root elements of the page are not getting removed. 

I checked and found that this root GridLayout has only one reference of RenderNode, which should be collected on FinalizerThread. But as it is a low priority thread, the views are being there on memory for some large time, which causes the OOM sometimes.

I have heapDumps for this scenarios with me. So if you need it let me know.

### Which platform(s) does your issue occur on?
Tested on Android device (ver: 6.0, heapSize: 256MB), emulator (ver: 5.0.1, heapSize: 32MB)

### Please provide the following version numbers that your issue occurs with:
- CLI: 2.2.1
- Cross-platform modules: 2.2.1
- Runtime(s): Android- 2.2.0

### Please tell us how to recreate the issue in as much detail as possible.
1. Run the sample application shared [here](https://github.com/shripalsoni04/ns-ng-gridview-memory-leak). 
2. Navigate between the two pages for around 5 times. (**NOTE**: Perform back navigation using Back Button on actionBar, as currently I have added code of calling GC() only there).
3. Open Android Monitor in Android Studio and select the device and the application process: org.nativescript.nsnggridviewmemoryleak
4. In memory panel, click on Initiate GC twice (Twice due to Finalizer objects in Android).
5. In memory panel, click on Dump Java Heap.
6. Once heapDump is opened, click on byte[] Class Name. Now in instance panel you can see that the image is not getting cleared and remaining on Heap. You can also check it on MAT for more details if you want. I have checked on both.
7. Now, in the application code, open app/pages/page2/page2.component.html.
8. Comment Case1 -> Memory Leak section and uncomment Case 1 -> No Memory Leak section.
9. Run the application and repeat steps from 2-6.
10. Similarly you can check the output of Case 2, which has multiple views instead of only image view to justify that the issue is not with ImageView but with GridView.

### Is there code involved? If so, please share the minimal amount of code needed to recreate the problem.
https://github.com/shripalsoni04/ns-ng-gridview-memory-leak

**Comment by tsonevn** [comment]
Hi @shripalsoni04,

This issue could be caused due to not clearing the Image properly as you mentioned in your issue. This is known issue for Android. As  a solution you could  use nativescript-fresco plugin for Android to load your Images, which will help to clear the memory, without memory leak. Example could be found [here](https://github.com/NativeScript/nativescript-fresco). and for angular [here](https://github.com/telerik/nativescript-ui-samples-angular/blob/release/sdkAngular/app/listview/item-layouts/listview-item-layouts-linear.component.android.html).

Regarding to that you should also register FrescoDrawee in your main.ts file as it shown [here](https://github.com/telerik/nativescript-ui-samples-angular/blob/release/sdkAngular/app/listview/item-layouts/listview-item-layouts-linear.component.android.html).

It would help if you could give us info about Android API level.

Regards,
@tsonevn

**Comment by shripalsoni04** [comment]
Thanks for your suggestion regarding nativescript-fresco.

I think the issue is not of ImageView. As described in the issue, in the sample app, when we use Image inside **StackLayout**, it is getting cleared from memory properly on back navigation. But when we use it inside **GridLayout**, it is not getting cleared from memory because GridLayout is not releasing the reference of its direct children. So probably the issue is with GridLayout.

As described in my issue comment, if you see code snippet 3 & 4, the issue is coming for Labels also. I used ImageView in code snippet 1 & 2 just to explain that there is memory leak, because it is easier to detect memory leak using Image as it consumes more memory than Labels. 

I did analysis of heapDumps In eclipse MAT and I have observed that the direct children of any GridLayout are staying there in memory. So if any of the GridLayout child is a GridLayout, it will keep its children also in memory. But if the direct child is a StackLayout, than only that StackLayout will be there in memory and its children will be removed.

I have tested this scenarios on  Android device (API level 23) and Emulator (API level 21). 
 
Just try to follow the 10 steps shown in issue description to reproduce this memory leak at your end.

Let me know if you want any other information.

Thanks.

**Comment by tsonevn** [comment]
Hi @shripalsoni04,
Thank you for your reply.
I reviewed your sample project, while  following the described steps. However the only way I was able to reproduce memory leak was when I load really large Image inside the ImageView. 
I have tested the given project on emulator and real devices and on both the same behaviour has been reproduced. 
Regarding to the problem, could you replace `ImageView` with `FrescoDrawee` and to verify whether this will solve this issue. You could also review the sample below.

page2.component.html
```
<ActionBar title="Title" icon="">
  <NavigationButton 
    text="Back" 
    android.systemIcon="ic_menu_back"
    (tap)="goBack()">
</NavigationButton>
</ActionBar>

<!-- ============================Case 1: Only one Image View============================ -->
<!-- Memory Leak -->
<GridLayout>
    <FrescoDrawee height="250" imageUri="https://unsplash.it/400/200?image=1067"></FrescoDrawee>
</GridLayout>

<!--  No Memory Leak -->
<!--<StackLayout>
    <FrescoDrawee height="250" imageUri="https://unsplash.it/400/200?image=1067"></FrescoDrawee>
</StackLayout>-->

<!-- ============================Case 2: With multiple views ============================ -->
<!-- Memory Leak -->
<!--<GridLayout rows="auto, auto, auto">
    <StackLayout row="0">
        <Label text="Label 1"></Label>
    </StackLayout>
    <Label row="1" text="Label 2"></Label>
    <StackLayout row="2">
        <FrescoDrawee height="250" imageUri="https://unsplash.it/400/200?image=1067"></FrescoDrawee>
    </StackLayout>
</GridLayout>-->

<!-- No Memory Leak -->
<!--<StackLayout>
    <GridLayout rows="auto, auto, auto">
        <StackLayout row="0">
            <Label text="Label 1"></Label>
        </StackLayout>
        <Label row="1" text="Label 2"></Label>
        <StackLayout row="2">
            <FrescoDrawee height="250" imageUri="https://unsplash.it/400/200?image=1067"></FrescoDrawee>
        </StackLayout>
    </GridLayout>
</StackLayout>-->
```

main.ts

```
import {nativeScriptBootstrap} from "nativescript-angular/application";

import * as elementRegistryModule from 'nativescript-angular/element-registry';
import {AppComponent} from "./app.component";
import { APP_ROUTER_PROVIDERS } from './app.routes';


import * as frescoModule from "nativescript-fresco";
import * as applicationModule from "application";

if (applicationModule.android) {
    applicationModule.onLaunch = function (intent) {
        frescoModule.initialize();
    };
}
elementRegistryModule.registerElement("FrescoDrawee", () => frescoModule.FrescoDrawee);

nativeScriptBootstrap(AppComponent, [APP_ROUTER_PROVIDERS]);
```

I hope this will solve your problem.

**Comment by shripalsoni04** [comment]
Hi @tsonevn 

Thanks for reviewing the sample code and providing code snippets of FrescoDrawee.

Actually, as explained in previous comment, I used ImageView just to show issue of memory leak in gridLayout. ( because generally image take much memory than any other view, so it is easier to detect memory leak when using image as compared to any other view element) 

But this memory leak issue is coming also when there is only label in gridLayout. To demonstrate this, I have added Case 3 to my [sample app](https://github.com/shripalsoni04/ns-ng-gridview-memory-leak) 

The HTML code for Case 3 is as follows:
```
<!-- Memory Leak -->
<GridLayout>
    <Label 
      textWrap="true" 
      [text]="longText">
    </Label>
</GridLayout>

<!-- No Memory Leak -->
<!--<StackLayout>
    <Label 
      textWrap="true" 
      [text]="longText">
    </Label>
</StackLayout>-->
```

So now, there is no image anywhere.

I took heap snapshots for both of the above cases by following 10 steps mentioned in my first comment. I navigated to the page with label 10 times. Below are the screenshots of heap dumps showing the issue.

## Label in StackLayout
When we use Label in StackLayout, all the TextViews are getting cleared, except one, which may be cleared when FinalizerThread run next time.

![label-in-stacklayout-dominator-tree](https://cloud.githubusercontent.com/assets/4411891/18548910/d792042a-7b68-11e6-8550-ce575131b264.png)

In below screenshot, it is path2gc for the TextView left. Note that, it is removed from StackLayout.

![label-in-stacklayout-path2gc]
(https://cloud.githubusercontent.com/assets/4411891/18548918/e21d055c-7b68-11e6-91db-395aae5bba8c.png)

## Label in GridLayout
When we use Label in GridLayout, we can see there are 10 TextViews left in the memory. They are not getting removed when we call GC any number of times.

![label-in-gridlayout-dominator-tree](https://cloud.githubusercontent.com/assets/4411891/18549001/681bd296-7b69-11e6-9828-b96664ac8744.png)

In below screenshot of path2gc of one of the TextView, note that the TextView is still tied to the GridLayout. 

![label-in-gridlayout-path2gc](https://cloud.githubusercontent.com/assets/4411891/18549032/9ac2ed7e-7b69-11e6-901e-2e05677da244.png)

I think the direct children of GridLayout are not getting released from the memory, irrespective of what kind of view it is and this results in continuous and slow increase in memory on navigation on Android.

I have tested the above Case 3, on both Android device (API level 23) and Emulator (API level 21).

You can try to run the updated sample app to check this issue at your end.

Hope this information will be helpful to you.

Thanks.

**Comment by tsonevn** [comment]
Hi @shripalsoni04,
Thank you for your further explanation.
After a little research I found that this is a real issue for Android and caused by not properly clearing GridLayout children from the memory. This problem has been reproduced on NativeScript version 2.2.1. Regarding to the this, we have memory and performance improvements  in our new NS version, which was released yesterday. Could you upgrade to version 2.3.0 and to verify whether you will have the same problem. [Upgrade instructions](https://docs.nativescript.org/releases/upgrade-instructions).

Please let me know if you still have this problem

**Comment by shripalsoni04** [comment]
Hello @tsonevn 

Thanks for confirming issue on 2.2.1

I upgraded to 2.3.0 version and tested the sample app with latest code on both Android device (API level 23) and Emulator (API level 21). I confirm that the issue is still there with latest version.

**Comment by tsonevn** [comment]
related to issue [#2741](https://github.com/NativeScript/NativeScript/issues/2741)

**Comment by shripalsoni04** [comment]
Hi @tsonevn ,

This issue is a different from #2741 .  In this issue, the memory leak occurs even when we use normal fade transition. If you check in my [sample app](https://github.com/shripalsoni04/ns-ng-gridview-memory-leak) that I have not used Slide transition in this sample app.
Also this issue of gridLayout memory leak occurs on both API 21 and 23 (Note: I haven't tested on other API Levels) irrespective of what transition type we use for navigation.

While in #2741, the issue occurs only when we use Slide transition on API 23. That is the problem with Slide Transition only and it is not related to GridLayout leak.

I request you to keep this issue open so that we can remove this bug from the framework.

Hope this will clarify the difference between both the issues #2712 and #2741.

Let me know in case of any confusion.

Thanks.

**Comment by tsonevn** [comment]
Hi @shripalsoni04,
I agree with you that the both issues could be different. In regarding to that could you verify whether you will have the same memory leak in non-NativeScript Anguler2 project.

Regards,
@tsonevn

**Comment by shripalsoni04** [comment]
Hi @tsonevn ,

I have created a [sample app](https://github.com/shripalsoni04/ns-gridLayout-memory-leak) in pure nativescript without using angular to show issue of memory leak in gridLayout.

I tested that on both API 21 and 23 and I confirm that the same issue is there in pure nativescript app also. 

To reproduce this at your end, just follow the 10 steps mentioned in main issue comment for this sample app.

Thanks.
